### PR TITLE
[sdk+hosting] Move AddOpenTelemetry & OpenTelemetryBuilder into Hosting package

### PR DESCRIPTION
**Pull request by CodeBlanch**
Alternative to #4151
Includes #4164

## Changes

* Moves `AddOpenTelemetry` extension and `OpenTelemetryBuilder` class into the hosting package.
* Removes `StartWithHost` extension.

## Details

This PR removes `StartWithHost` by moving the `IServiceCollection` configuration pattern into the hosting package. Users who want to use the `IServiceCollection` startup pattern will need the hosting dependency.

## TODOs

* [X] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [X] Fix tests
* [ ] Changes in public API reviewed

(Will do documentation updates as a follow-up PR.)

### CODE REVIEW

File: src/OpenTelemetry/AssemblyInfo.cs [hunk 1]
```diff
@@ -21,5 +21,6 @@
 [assembly: InternalsVisibleTo("OpenTelemetry.Exporter.Prometheus.AspNetCore" + AssemblyInfo.PublicKey)]
 [assembly: InternalsVisibleTo("OpenTelemetry.Exporter.Prometheus.AspNetCore.Tests" + AssemblyInfo.PublicKey)]
 [assembly: InternalsVisibleTo("OpenTelemetry.Exporter.Prometheus.HttpListener.Tests" + AssemblyInfo.PublicKey)]
+[assembly: InternalsVisibleTo("OpenTelemetry.Extensions.Hosting" + AssemblyInfo.PublicKey)]
```
Comment (CodeBlanch) [inline src/OpenTelemetry/AssemblyInfo.cs:24]: @alanwest and I discussed keeping `OpenTelemetryBuilder` inside the SDK. I decided to move it to hosting because it wasn't used inside SDK at all after moving `AddOpenTelemetry`. We think there might be a better API we can ship for the `Sdk.Create*` style which also uses  `OpenTelemetryBuilder`. What I'm thinking is in vNext (or whatever) we can explore that. We can always move `OpenTelemetryBuilder` back into SDK via [TypeForwardedToAttribute](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.typeforwardedtoattribute). If we do that, then we can also remove this `InternalsVisibleTo` cheat, it is only needed because `OpenTelemetryBuilder` has to register some SDK internals.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs [hunk 2]
```diff
@@ -27,6 +30,29 @@ namespace Microsoft.Extensions.DependencyInjection;
...
+    /// </remarks>
+    /// <param name="services"><see cref="IServiceCollection"/>.</param>
+    /// <returns>The supplied <see cref="OpenTelemetryBuilder"/> for chaining
+    /// calls.</returns>
+    public static OpenTelemetryBuilder AddOpenTelemetry(this IServiceCollection services)
+    {
+        Guard.ThrowIfNull(services);
+
+        services.TryAddEnumerable(
+            ServiceDescriptor.Singleton<IHostedService, TelemetryHostedService>());
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:51]: All this effort for this one dumb line 🤣

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetry.Extensions.Hosting.csproj [hunk 3]
```diff
@@ -1,13 +1,11 @@
...
   <PropertyGroup>
     <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
     <TargetFrameworks>netstandard2.0</TargetFrameworks>
     <Description>Contains extensions to start OpenTelemetry in applications using Microsoft.Extensions.Hosting</Description>
     <RootNamespace>OpenTelemetry</RootNamespace>
     <MinVerTagPrefix>core-</MinVerTagPrefix>
-
-    <!-- this is temporary. will remove in future PR. -->
-    <Nullable>disable</Nullable>
+    <Nullable>enable</Nullable>
```
Comment (TimothyMothra) [inline src/OpenTelemetry.Extensions.Hosting/OpenTelemetry.Extensions.Hosting.csproj:8]: Would it make sense to move this into a separate PR?
Just thinking it might simplify life. :)

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs (hunk 2 above)
Comment (TimothyMothra) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:51]: Do we need a log here if the TryAdd fails?

File: src/OpenTelemetry/AssemblyInfo.cs (hunk 1 above)
Comment (alanwest) [reply src/OpenTelemetry/AssemblyInfo.cs:24]: The other thing we discussed was whether the `AddOpenTelemetry(this IServiceCollection)` extension had any utility outside of a hosting use case. If we ever thought this might be useful, forwarding this extension might be trickier, no?

Though, it seems some folk find the non-hosting use case unlikely, so I suppose in that case should we just not concern ourselves with it?

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:51]: `TryAddEnumerable` (oddly) does not return true or false so we have no way of knowing. But even then, it only no-ops if an exact march for service type (IHostedService) and implementation type (TelemetryHostedService) is found. So it would only no-op for repeated calls to `AddOpenTelemetry` which is totally fine.

File: src/OpenTelemetry.Extensions.Hosting/OpenTelemetry.Extensions.Hosting.csproj (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Extensions.Hosting/OpenTelemetry.Extensions.Hosting.csproj:8]: Fair point. I had to move some code in from SDK which was nullable enabled so I had to either clobber that and throw away usefulness or update the remaining things to be enabled. Nothing has been simple on this! 🤣

File: src/OpenTelemetry/AssemblyInfo.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/AssemblyInfo.cs:24]: I can't think of any strong use case for needing to call `AddOpenTelemetry` beyond setting up your host. The main case for acting on the `IServiceCollection` beyond host setup is the library author case, but we have `services.ConfigureTracerProviderBuilder` & `services.ConfigureMeterProviderBuilder` for that.

If we found a use case and wanted to move `AddOpenTelemetry` back into SDK, you can move a type. So in this case `OpenTelemetryServicesExtensions`. It currently has some other obsolete methods, but the plan was to drop them before stable so if we do that, I think the potential to move it is there. But if we did move it we would be right back in the hole of `IHostedService` dependency not being part of SDK.

Comment (alanwest) [reply src/OpenTelemetry/AssemblyInfo.cs:24]: > But if we did move it we would be right back in the hole of IHostedService dependency not being part of SDK.

Right, this is the tricky part. Unless M.E.DI.A could introduce that hook for when the provider has been built. That said, I'm not really concerned since you're thinking AddOpenTelemetry goes hand-in-hand with a hosted use case.

Also, 👍 to:

> but we have services.ConfigureTracerProviderBuilder & services.ConfigureMeterProviderBuilder for that.

Comment (CodeBlanch) [reply src/OpenTelemetry/AssemblyInfo.cs:24]: The reason I reverted #4151 and went this direction was that whole design was built on the hope that we could do something in .NET 8 to M.E.DI.A to supplant the need for `IHostedService`. I took a stab at making that so, and couldn't see a viable path to do it which wouldn't be breaking to custom `IServiceProvider`s 😢 Also @noahfalk's feedback was the separation between "dropping services into a collection" and "starting up/requesting services" was intentional in the design.

**Review by cijothomas — APPROVED** [review]

**Review by alanwest — APPROVED** [review]
